espnet2/enh/layers/iva.py

```diff
@@ -0,0 +1,513 @@
+
+    else:
+        # Iterative Source Steering updates
+        X, W, A = spatial_model_update_iss(X, weights, W=W, A=A, eps=eps)
+
+    return X, W, A
+
+
+def auxiva_iss_one_step_dmc(W, A, model, eps, two_chan_ip2, X_original):
+    """
+    wrapper that does not take the separated signal as input
+    to use with demixing matrix checkpointing
+    """
+    X = demix(W, X_original)
+    X, W, A = auxiva_iss(X, W, A, model, eps, two_chan_ip2, X_original)
+    return W, A
+
+
+def auxiva_iss(
+    X: torch.Tensor,
+    n_iter: Optional[int] = 20,
+    n_src: Optional[int] = None,
+    model: Optional[callable] = None,
+    eps: Optional[float] = 1e-6,
+    two_chan_ip2: Optional[bool] = True,
+    proj_back_mic: Optional[bool] = 0,
+    use_dmc: Optional[bool] = False,
+    checkpoints_iter: Optional[List[int]] = None,
+    checkpoints_list: Optional[List] = None,
+) -> torch.Tensor:
+
+    """
+    Blind source separation based on independent vector analysis with
+    alternating updates of the mixing vectors
+
+    Parameters
+    ----------
+    X: Tensor, shape (..., n_channels, n_frequencies, n_frames)
+        STFT representation of the signal
+    n_iter: int, optional
+        The number of iterations (default 20)
+    n_src: int, optional
+        When n_src is less than the number of channels, the n_src most energetic
+        are selected for the output
+    model: SourceModel
+        The model of source distribution (default: Laplace)
+    eps: float
+        A small constant to make divisions and the like numerically stable
+    two_chan_ip2: bool
+        For the 2 channel case, use the more efficient IP2 algorithm (default: True).
+        Ignored when using more than 2 channels.
+    proj_back_mic: int
+        The microphone index to use as a reference when adjusting the scale and delay
+    use_dmc: bool, optional
+        If True, use checkpointing of the demixing matrix to save memory
+    checkpoints_iter: List of int
+        Optionally, we can keep intermediate results for later analysis
+        Should be used together with checkpoints_list
+    checkpoints_list: List
+        An empty list can be passed and the intermediate signals are
+        appended to the list for iterations numbers contained in checkpoints_iter
+
+    Returns
+    -------
+    X: Tensor, shape (..., n_channels, n_frequencies, n_frames)
+        STFT representation of the signal after separation
+    """
+
+    n_chan, n_freq, n_frames = X.shape[-3:]
+
+    if model is None:
+        model = LaplaceModel()
+
+    # for now, only supports determined case
+    assert callable(model)
+
+    # keep track of original input signal
+    X_original = X.clone()
+
+    if proj_back_mic is not None or use_dmc:
+        assert (
+            0 <= proj_back_mic < n_chan
+        ), "The reference microphone index must be between 0 and # channels - 1."
+        W = X.new_zeros(X.shape[:-3] + (n_freq, n_chan, n_chan))
```

Comment:
For better compatibility with older PyTorch versions, it it better to replace this line with
```python
        W = X.new_full(X.shape[:-3] + (n_freq, n_chan, n_chan), 0)
```

> This is because `ComplexTensor` currently does not have the method `new_zeros`.